internal/api/token_refresh.go

```diff
@@ -172,6 +181,27 @@ func (a *API) RefreshTokenGrant(ctx context.Context, w http.ResponseWriter, r *h
 				return internalServerError("error generating jwt token").WithInternalError(terr)
 			}
 
+			refreshedAt := a.Now()
+			session.RefreshedAt = &refreshedAt
+
+			userAgent := r.Header.Get("User-Agent")
+			if userAgent != "" {
+				session.UserAgent = &userAgent
+			} else {
+				session.UserAgent = nil
+			}
+
+			ipAddress := utilities.GetIPAddress(r)
+			if ipAddress != "" {
+				session.IP = &ipAddress
+			} else {
+				session.IP = nil
+			}
```

Comment:
does it make sense to create a `GrantParams` struct here and use `grantParams.FillFromRequest` instead?

Reply:
I actually tried to do that, but the whole grant-params flow is so complicated I decided to just update them in-place. The biggest issue is the latest code where we return an already existing refresh token and that case gets super hairy to fix via grant-params.